Codeinwp/neve
# [DEV] Update NPM packages and JS build

### Summary


1. Updated NPM packages to their latest version 🚀 
2. Using `wp-scripts` built-in build for JS files instead of a custom one with `ts-loader`.
3. Update the Rollup config files. File renamed to `.mjs` and `sourceMap` to `sourcemap`.
4. The build can be run with Node 16, but Node 18 is the recommended way.
5. Remove unused libs like `ts-loader`, `css-loader`, etc.

ℹ️ The newer version of Eslint is more strict, so we have new warnings like:

```
/Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/assets/apps/customizer-controls/src/rich-text/RichTextComponent.js
  49:5  warning  React Hook useEffect has missing dependencies: 'isVisible' and 'section'. Either include them or remove the dependency array                     react-hooks/exhaustive-deps
  58:5  warning  React Hook useEffect has missing dependencies: 'controlId', 'isVisible', and 'updateValues'. Either include them or remove the dependency array  react-hooks/exhaustive-deps

/Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/assets/apps/customizer-controls/src/spacing/SpacingComponent.js
  102:5  warning  React Hook useEffect has missing dependencies: 'control.id', 'defaultValue', 'dependsOn', and 'updateControlValue'. Either include them or remove the dependency array  react-hooks/exhaustive-deps

/Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/assets/apps/customizer-controls/src/toggle/ToggleComponent.js
  17:5  warning  React Hook useEffect has missing dependencies: 'control.id', 'toggleValue', and 'value'. Either include them or remove the dependency array  react-hooks/exhaustive-deps
```

#### Build Optimizations 

When new versions change, this change has slightly increased the build size, even though some things are much smaller.

For CSS, the last version of the SASS compiler can have some smart replacements like `:nth-child(even)` to `:nth-child(2n)`, which saves some space. It is stuck at some basic things like removing the space in this structure: `var(--a, var(--b))`, which adds up and thus makes it cross the limit. 

Fortunately, with a new version comes new opportunities. With CSS minifier, we can use level 2 optimization to make some smart arrangements in the selector to reuse more code. This option enabled the size to be reduced to `1kb` 👌 

It is a big win for CSS, but not for JS...

For the JS part, it was a bit sad of a change; the new Babel version has new tricks to deal with JS, and those new features increased the size of the scripts:

```
Old Babel Version
  proposal-numeric-separator { "ios":"12.2" }
  proposal-logical-assignment-operators { "firefox":"78", "ios":"12.2", "opera":"88", "samsung":"17" }
  proposal-nullish-coalescing-operator { "ios":"12.2" }
  proposal-optional-chaining { "ios":"12.2" }
  syntax-json-strings { "android":"103", "chrome":"102", "edge":"103", "firefox":"78", "ios":"12.2", "opera":"88", "safari":"15.5", "samsung":"17" }
  syntax-optional-catch-binding { "android":"103", "chrome":"102", "edge":"103", "firefox":"78", "ios":"12.2", "opera":"88", "safari":"15.5", "samsung":"17" }
  syntax-async-generators { "android":"103", "chrome":"102", "edge":"103", "firefox":"78", "ios":"12.2", "opera":"88", "safari":"15.5", "samsung":"17" }
  syntax-object-rest-spread { "android":"103", "chrome":"102", "edge":"103", "firefox":"78", "ios":"12.2", "opera":"88", "safari":"15.5", "samsung":"17" }
  transform-template-literals { "ios":"12.2" }
  proposal-export-namespace-from { "firefox":"78", "ios":"12.2", "safari":"15.5" }
  syntax-dynamic-import { "android":"103", "chrome":"102", "edge":"103", "firefox":"78", "ios":"12.2", "opera":"88", "safari":"15.5", "samsung":"17" }
  
 New Babel Version:
  transform-unicode-sets-regex { chrome < 112, firefox < 116, ios, opera_mobile < 75, safari < tp, samsung }
  proposal-class-static-block { ios < 16.4, safari < 16.4 }
  syntax-private-property-in-object
  syntax-class-properties
  syntax-numeric-separator
  syntax-nullish-coalescing-operator
  syntax-optional-chaining
  syntax-json-strings
  syntax-optional-catch-binding
  syntax-async-generators
  syntax-object-rest-spread
  syntax-export-namespace-from
  syntax-dynamic-import
  syntax-top-level-await
  syntax-import-meta 
```
As you can see, a lot of functionality in the old Babel version was at the proposal stage. And because of this, the new code slightly differs from the old one.

In the Rollup configuration file, you will see that I disabled: `'@babel/plugin-transform-parameters'` which does [this](https://babeljs.io/docs/babel-plugin-transform-parameters) -- the change was not present in the Old version.

For `frontend.js`, the limit was above with under 100B, so I put a little trick like making an alias for `document` to reduce the size. It worked 🔧 . But for `shop.js`, the trick was not possible since most of the code is from a third-party lib, and the new Babel version introduced some additional vars and increased the size with `234 B` -- for this, I think we need to increase the limit 😢 

#### ⚠️ Yarn changed with NPM

The new version of Eslint is not working with Yarn. It gives an error like:

```
yarn run lint
yarn run v1.22.21
$ npm-run-all lint:*
$ wp-scripts lint-js ./assets/js/src/
There was a problem loading formatter: /Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/node_modules/eslint/lib/cli-engine/formatters/stylish
Error: require() of ES Module /Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/node_modules/strip-ansi/index.js from /Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/node_modules/eslint/lib/cli-engine/formatters/stylish.js not supported.
Instead change the require of index.js in /Users/robert/Desktop/sites/dev/app/public/wp-content/themes/neve/node_modules/eslint/lib/cli-engine/formatters/stylish.js to a dynamic import() which is available in all CommonJS modules.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "lint:global" exited with 2.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

So I had to change to NPM, which gives proper package support. 

### Will affect visual aspect of the product

NO

### Screenshots 

### Test instructions


Since it is a big update on packages, there is a possibility for broken packages. Also, we need to keep in mind the compatibility. 

1. Make two instances, one with the latest version of WP and another with WP5.5 (our minimum). Also, install Neve Pro.
2. Load the version from the PR and check for errors. 

⚠️ When encountering an error, please post a screenshot an share the instance. We need to double-check to see if it is a new error introduced in this or if it also happened in the current version of Neve, and we just found out about it.

## Check before Pull Request is ready:



Closes #4143
